File: modules/ui-web/src/api/types/surface.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Surface Manifest types — a **composing barrel** over the generated wire projection.
 *
 * Tempdoc 884 (ADR-0038 amendment) retired the hand-written mirror that used to live here.
 * The authority is now `SSOT/schemas/surface.v1.json`, emitted from
 * `modules/app-agent-api/.../registry/Surface.java` by `SubstrateSchemaGenTest` and projected to
 * TypeScript by `scripts/codegen/gen-wire-schema-types.mjs` (registered in
 * `governance/contract-surfaces.v1.json` as `SurfaceWire`). Regenerate with
 * `./gradlew.bat :modules:app-api:updateSchemas` then
 * `node scripts/codegen/gen-wire-schema-types.mjs`; `check-wire-schema-types-regen` fails on drift.
 *
 * This file is not a mirror and must not become one again. Everything below is one of:
 *  1. a re-export or projection of `SurfaceWire` (the field SET comes from Java, mechanically);
 *  2. a **client-side-only** concern the backend never sends — `factory` (478 §4.A dispatch token,
 *     minted by the catalog at boot) and `splitPairing` (521 §22 Phase D, a plugin contribution
 *     merged client-side; it has zero Java source);
 *  3. a closed vocabulary list (`AUDIENCES`, `PLACEMENTS`) that exists as a runtime value, which a
 *     type projection cannot provide.
 *
 * Per slice 449 §0 D1: Surface is a **Manifest** (composition over primitives), not a fifth
 * primitive. Its catalog is parallel-but-distinct from the per-primitive catalogs
 * (Resource / Operation / Prompt / DiagnosticChannel). Same pattern as `registry.ts`'s
 * `ResourceWire` / `OperationWire` barrels.
 */

import { z } from 'zod';
import type { Presentation, Provenance } from './registry.js';
import type { Altitude } from '../generated/registry-enums.generated.js';
import type { SurfaceWire } from '../generated/schema-types/surface.js';
import { surfaceWireSchema } from '../generated/schema-types/surface.js';
export { surfaceWireSchema };
export type { SurfaceWire };

export type { Altitude };

// ============================================================
// Discriminator vocabularies — projected from the wire
// ============================================================

/**
 * Access-control audience axis, projected from `SurfaceWire`.
 *
 * Trust ordering for the audience-composition rule (slice 449 §0 D2):
 * `USER < OPERATOR < DEVELOPER`. AGENT is excluded from the comparison —
 * agent surfaces are consumed by headless tool APIs, not human chrome.
 */
export type Audience = SurfaceWire['audience'];

/** Closed list useful for exhaustive switches and test fixtures. */
export const AUDIENCES: readonly Audience[] = [
  'USER',
  'AGENT',
  'OPERATOR',
  'DEVELOPER',
] as const;

/**
 * Chrome-zone placement axis, projected from `SurfaceWire`.
 *
 * Adding a value is a Java change (`Placement.java`), not an edit here: the union is generated,
 * so a new zone appears in this type automatically. A new zone's landmark-role mapping
 * (shell-v0/display/landmarks.ts) is a review-tier obligation (tempdoc 930 chunk H retired the
 * static `check-a11y-closure` gate that used to assert this by parsing the list below).
 */
export type Placement = SurfaceWire['placement'];

/**
 * Closed list useful for exhaustive switches.
 *
 * This is a runtime VALUE, which is why it is hand-maintained rather than projected.
 *
 * The `readonly Placement[]` annotation checks only that every entry IS a `Placement` — a subset
 * type-checks fine, so it cannot catch a dropped or newly-added zone. `PLACEMENT_CLOSURE` below is
 * what makes that a compile error.
 */
export const PLACEMENTS: readonly Placement[] = [
  'COMMAND',
  'RAIL',
  'STAGE',
  'HUD',
  'STATUS',
  'DRAWER',
  'MODAL',
  'DEEPLINK',
  'HEADLESS_AGENT_TOOL',
] as const;

/**
 * Compile-time closure over the generated `Placement` union: every member must appear as a key.
 * Adding a zone in Java (`Placement.java`) regenerates the union, and this declaration then fails to
 * type-check until the new zone is listed here AND in `PLACEMENTS` above — which is what forces the
 * a11y gate to see it and demand a landmark-role mapping.
 *
 * Type-level only; `void`-consumed so it is not an unused binding. It is deliberately NOT exported:
 * the registry barrel may not declare a second hand-authored wire shape (`contribution-surface`
 * gate), and this is an assertion, not a shape.
 */
const PLACEMENT_CLOSURE: Record<Placement, true> = {
  COMMAND: true,
  RAIL: true,
  STAGE: true,
  HUD: true,
  STATUS: true,
  DRAWER: true,
  MODAL: true,
  DEEPLINK: true,
  HEADLESS_AGENT_TOOL: true,
};
void PLACEMENT_CLOSURE;

// ============================================================
// Common value types
// ============================================================

/**
 * Surface id. Serialized as a bare string (`SurfaceRef.java`'s `@JsonValue`).
 * Pattern: `core.<name>` or `vendor.<plugin-id>.<name>` — the generated schema carries the regex.
 */
export type SurfaceRef = SurfaceWire['id'];

/**
 * Typed cross-reference graph from a Surface to the primitives it consumes — a direct projection of
 * the wire's `consumes` object.
 *
 * `SurfaceConsumes.java` implements `PreciseWire` and its compact constructor defaults every `Set`
 * to `Set.of()`, so the generated schema marks all five graphs `required` and non-null. Nothing is
 * restated here: the field set, its cardinality and its nullability all come from Java
 * mechanically, so a renamed or removed component changes this type rather than silently diverging
 * from it.
 *
 * Slice 491 §9.D Phase E (C0): `conversationShapes` joins the graph. Surfaces hosting a chat shape
 * (e.g. `core.agent-surface` consumes `core.agent`) declare it here, and `<jf-chat-shape-mount>`
 * reads it to resolve which shape's view factory to instantiate. It is the ONE key relaxed to
 * optional at the FE boundary — client-side Surface literals (plugin contributions, test fixtures)
 * predate Phase E and omit it, even though the wire always carries it. Every other key comes
 * through `Omit` untouched, so the required set is derived by exclusion rather than restated.
 */
export type SurfaceConsumes = Omit<SurfaceWire['consumes'], 'conversationShapes'> &
  Partial<Pick<SurfaceWire['consumes'], 'conversationShapes'>>;

// ============================================================
// Surface manifest
// ============================================================

/**
 * A Surface catalog entry as the FE holds it: the generated wire shape, plus the two client-side
 * fields the backend never sends.
 *
 * `altitude` and `members` are relaxed to optional because the FE constructs Surface objects that
 * predate them (`SurfaceCatalogClient` merge paths, plugin contributions, test fixtures); the wire
 * always provides both, and an absent `altitude` is read as `PRODUCT`, mirroring the Java record's
 * `null → PRODUCT` default. `riskTier` and `stateSchema` are backend concerns that ride the same
 * wire; they are optional here for the same reason.
 */
export type Surface = Omit<
  SurfaceWire,
  | 'presentation'
  | 'provenance'
  | 'consumes'
  | 'altitude'
  | 'members'
  | 'riskTier'
  | 'stateSchema'
> & {
  presentation: Presentation;
  provenance: Provenance;
  consumes: SurfaceConsumes;
  /**
   * Tempdoc 571 — the surface's **altitude**: the governing axis that determines its home (rail
   * band) and its core-vs-plugin eligibility, as a projection of the primary authority it carries.
   * The authoritative declaration and the foreclosures (`TRUST ⟹ CORE`;
   * `channel ⟹ DIAGNOSTIC`) live in the Java `CoreSurfaceCatalog` + the `surface-altitude` gate.
   */
  altitude?: Altitude;
  /**
   * Tempdoc 571 §11 / 578 — the declared host/member composition relationship. A host surface names
   * the member surfaces it presents inside itself (e.g. System hosts Health/Logs/Activity).
   * Membership is the single home-authority: a member is excluded from the rail and its deep-link
   * resolves to the host. Order is the declared tab order. Absent/empty ⇒ this surface hosts
   * nothing. The `surface-composition` gate enforces one-home integrity.
   */
  members?: SurfaceRef[];
  /** Backend-only navigation risk tier (tempdoc 550 WA-4); on the wire, unused by the FE. */
  riskTier?: SurfaceWire['riskTier'];
  /** Backend-declared surface state schema; on the wire, unused by the FE. */
  stateSchema?: SurfaceWire['stateSchema'];
  /**
   * Tempdoc 521 §22 Phase D — declarative split pairing. When this surface is the primary pane of
   * a split-stage layout and the user has not chosen a secondary surface explicitly, `Shell` reads
   * `splitPairing.secondary` as the curated default. Plugins can declare their preferred pair
   * without touching host code. Absent ⇒ no curated pair; fallback to "first non-primary rail
   * surface."
   *
   * **Client-side only** — no Java counterpart exists; it is merged in by
   * `mergePluginSurfaceContributions`, never served.
   */
  splitPairing?: {
    secondary: string;
  };
  /**
   * 478 §4.A — SurfaceFactory minted by the FE catalog at boot/merge time. The factory captures the
   * validated mountTag in a closure; consumers (Stage.render()) call factory.mount() to construct
   * the surface element WITHOUT going through template-string interpolation.
   *
   * **Client-side only** — wire-shape Surfaces arrive with `factory` undefined and the catalog
   * stamps it on receipt. Stage falls back to the legacy `mountTag` + customElements.get path when
   * it is absent (transitional; V1.5.2 marks factory required).
   */
  factory?: SurfaceFactory;
};

/**
 * 478 §4.A — opaque dispatch token. The catalog is the only mint-
 * site; consumers cannot construct a SurfaceFactory because they
 * cannot reach the module-private WeakSet of valid factories.
 *
 * Two-layer brand model (post-reviewer-pass critical-analysis):
 *
 * 1. **Type-level brand** — the `__dispatchBrand: symbol` field
 *    prevents accidental constructions like `{mount: () => el}`
 *    from satisfying the SurfaceFactory type. Consumers can READ
 *    the symbol off a real factory but the type system rejects
 *    objects without the field.
 *
 * 2. **Runtime brand verification** — the catalog's
 *    `mountSurface()` helper checks WeakSet membership. The
 *    WeakSet is module-private inside SurfaceCatalogClient; a
 *    consumer with only a factory reference cannot insert into
 *    it. So stealing `__dispatchBrand` from a real factory and
 *    forging `{__dispatchBrand: stolen, mount: () => evilEl}`
 *    passes the type check (Symbol matches) but FAILS the
 *    runtime verification (forged factory not in WeakSet).
 *
 * The factory's mount() returns an HTMLElement constructed from
 * the registered custom element. customElements.get() validates
 * the tag is a properly-registered name (HTML spec PCEN regex);
 * this validation is encapsulated inside the factory closure
 * rather than exposed to consumers.
 */
export interface SurfaceFactory {
  /** Type-level brand — see SurfaceFactory doc comment. */
  readonly __dispatchBrand: symbol;
  /**
   * Mount the surface as a fresh HTMLElement. Throws if the
   * registered class can't be constructed (degenerate edge case;
   * customElements.get returned a class but `new klass()` threw).
   */
  mount(opts?: { apiBase?: string; host_?: unknown }): HTMLElement;
}

/**
 * Catalog envelope returned by `GET /api/registry/surfaces`.
 * Wire shape mirrors `RegistryController.handleSurfaces`.
 */
export interface SurfaceCatalog {
  $schema?: string;
  schemaVersion: string;
  catalogVersion: number;
  namespace: string;
  primitive: 'Surface';
  entries: Surface[];
}

/**
 * Runtime validator for the Surface catalog envelope (tempdoc 884; the 560 §4c Phase B parse
 * boundary, extended to the Surface Manifest). Each entry is validated by the GENERATED
 * `surfaceWireSchema` — the single runtime authority for the Surface wire shape.
 */
export const surfaceCatalogSchema = z.object({
  $schema: z.string().optional(),
  schemaVersion: z.string(),
  catalogVersion: z.number(),
  namespace: z.string(),
  primitive: z.literal('Surface'),
  entries: z.array(surfaceWireSchema),
});

File: modules/app-agent-api/src/main/java/io/justsearch/agent/api/registry/Surface.java
/* SPDX-License-Identifier: Apache-2.0 */
package io.justsearch.agent.api.registry;

import java.util.List;
import java.util.Objects;
import java.util.Optional;

/**
 * Chrome-surface manifest — a structured composition over existing primitives.
 *
 * <p>Per slice 449 §0 D1: Surface is a {@code Manifest} (the second Manifest
 * tier alongside Plugin), <strong>not</strong> a fifth primitive. Manifests
 * compose primitives into shippable units; primitives are the BE-declared
 * truth Manifests reference. The framework's primitive count stays at four
 * (Operation / Resource / Prompt / DiagnosticChannel); the Manifest count
 * grows from one (Plugin, implicit) to two (Plugin + Surface, explicit) when
 * slice 449 ships.
 *
 * <p>Surface does NOT implement {@link RegistryEntry} — that interface is the
 * sealed marker for the four primitives. Surface is its own kind of
 * registered metadata; its catalog ({@link SurfaceCatalog}) is parallel-but-
 * distinct from the per-primitive catalogs.
 *
 * <h3>Field guidance</h3>
 *
 * <ul>
 *   <li>{@link #id}: namespaced identifier; mirrors {@link OperationRef} format.
 *       See {@link SurfaceRef}.
 *   <li>{@link #presentation}: i18n-keyed labelKey + descriptionKey. Resolves
 *       through {@code /api/messages/registry-surface/{locale}}.
 *   <li>{@link #audience}: declared access-control audience. The
 *       {@link io.justsearch.app.services.registry.validator.SurfaceAreaValidator}
 *       computes the {@code effectiveAudience} per slice 449 §0 D2 and
 *       rejects the entry if {@code audience < effectiveAudience}. The
 *       chrome's {@code visibleAudienceSet} (a chrome-level state) controls
 *       which audience entries are rendered.
 *   <li>{@link #placement}: chrome-zone placement. Each chrome zone's
 *       dispatcher filters by {@code placement = ZONE && audience ∈ visibleAudienceSet}.
 *   <li>{@link #consumes}: typed cross-reference graph to backing primitives.
 *       The validator follows these references for audience composition
 *       (DiagnosticChannel.consumerPermission lifts the floor) and (V1.5)
 *       cross-reference enforcement (slice 3a-1-8c).
 *   <li>{@link #mountTag}: custom-element name the chrome dispatches to.
 *       Side-effect import registers the element; the dispatcher just mounts
 *       the resolved tag.
 *   <li>{@link #provenance}: CORE / TRUSTED_PLUGIN / UNTRUSTED_PLUGIN. The
 *       provenance tier is the audience floor for plugin-contributed
 *       surfaces per the §0 D2 rule.
 *   <li>{@link #stateSchema}: per slice 489 §5 — optional typed contract
 *       declaring the addressable state shape (query, filters, scroll, etc.)
 *       this surface accepts and exposes. {@link Optional#empty()} means the
 *       surface is not URL-addressable; activation falls back to a default
 *       view. Per slice 489 §17.2 resolution this field is optional, not
 *       mandatory, during the substrate's initial rollout.
 *   <li>{@link #altitude}: the governing axis (tempdoc 571) — determines the surface's home (rail
 *       band) and its core-vs-plugin eligibility, as a projection of the primary authority it
 *       carries. Declared per surface and witnessed by the {@code surface-altitude} gate
 *       ({@code TRUST ⟹ CORE}; {@code consumes-a-channel ⟹ DIAGNOSTIC}). {@link Altitude#PRODUCT}
 *       is the benign default. See {@link Altitude}.
 *   <li>{@link #members}: tempdoc 571 §11 / 578 — the declared host/member composition relationship.
 *       A host surface names the member surfaces it presents inside itself (e.g. System Health hosts
 *       Logs). Membership is the single home-authority: a member's home is its host, so it is
 *       excluded from the rail and its deep-link resolves to the host (read as-is, never derived). An
 *       empty list (the default) means the surface hosts nothing. <strong>Invariant (578 §10 U1): this
 *       field is NOT an input to {@link SurfaceAltitude#derive} — a host does not inherit or aggregate
 *       its members' altitudes; each surface (host and member) derives its own altitude from its own
 *       {@code consumes}.</strong> The {@code surface-composition} gate enforces one-home integrity
 *       (a member may not also be a RAIL surface, nor be hosted twice).
 * </ul>
 */
public record Surface(
    SurfaceRef id,
    Presentation presentation,
    Audience audience,
    Placement placement,
    SurfaceConsumes consumes,
    String mountTag,
    Provenance provenance,
    Optional<SurfaceStateSchema> stateSchema,
    RiskTier riskTier,
    Altitude altitude,
    List<SurfaceRef> members) implements Provenanced {

  public Surface {
    Objects.requireNonNull(id, "id");
    Objects.requireNonNull(presentation, "presentation");
    Objects.requireNonNull(audience, "audience");
    Objects.requireNonNull(placement, "placement");
    Objects.requireNonNull(mountTag, "mountTag");
    Objects.requireNonNull(provenance, "provenance");
    Objects.requireNonNull(stateSchema, "stateSchema");
    // Tempdoc 550 WA-4: navigating TO a surface traverses the same (SourceTier × RiskTier)
    // lattice as an Operation invocation, so "all actions traverse the spine" (538) holds
    // for Navigation too. Default LOW → Navigation × any tier = AUTO (no behavior change);
    // only a surface that exposes destructive state declares higher. Backend-only (read by
    // BackendIntentRouter.forwardNavigation); not part of the FE surface wire payload.
    riskTier = riskTier == null ? RiskTier.LOW : riskTier;
    // Tempdoc 571: altitude is the governing axis (home + core-vs-plugin eligibility). PRODUCT is the
    // benign default; DIAGNOSTIC / TRUST are declared explicitly and witnessed by the surface-altitude
    // gate (TRUST ⟹ CORE provenance; consumes-a-channel ⟹ DIAGNOSTIC). See {@link Altitude}.
    altitude = altitude == null ? Altitude.PRODUCT : altitude;
    consumes = consumes == null ? SurfaceConsumes.empty() : consumes;
    // Tempdoc 571 §11 / 578: the declared host/member relationship. Defaults to no members.
    // Read as-is (never derived); excluded from SurfaceAltitude.derive by design (578 §10 U1).
    members = members == null ? List.of() : List.copyOf(members);
    if (mountTag.isBlank()) {
      throw new IllegalArgumentException(
          "Surface " + id.value() + " mountTag must be non-blank");
    }
    // Structural placement-vs-audience invariants (the SurfaceAreaValidator
    // enforces the audience-composition rule; these compact-constructor
    // checks are the local invariants the record itself can guarantee).
    if (placement == Placement.HEADLESS_AGENT_TOOL && audience != Audience.AGENT) {
      throw new IllegalArgumentException(
          "Surface " + id.value() + " has placement HEADLESS_AGENT_TOOL but audience "
              + audience + "; HEADLESS_AGENT_TOOL placement requires AGENT audience");
    }
    if (audience == Audience.AGENT && placement != Placement.HEADLESS_AGENT_TOOL) {
      throw new IllegalArgumentException(
          "Surface " + id.value() + " has audience AGENT but placement " + placement
              + "; AGENT audience requires HEADLESS_AGENT_TOOL placement");
    }
  }

  /**
   * Returns a copy of this Surface with {@code altitude} set — the injection point for the DERIVED
   * altitude (tempdoc 571 §4c). {@code RegistryController} computes the altitude from {@code consumes}
   * + the consumed Resources' {@link Role}s and stamps it onto the wire via this wither, so the
   * surfaces wire carries the derived value rather than a hand-declared one.
   */
  public Surface withAltitude(Altitude newAltitude) {
    return new Surface(
        id, presentation, audience, placement, consumes, mountTag, provenance, stateSchema, riskTier,
        newAltitude, members);
  }

  /**
   * Returns a copy of this Surface declaring {@code newMembers} as its hosted member surfaces
   * (tempdoc 571 §11 / 578). The ergonomic way for a catalog to declare a host without expanding its
   * constructor call to the full positional form — mirrors {@link #withAltitude}. Membership is the
   * single home-authority (members are excluded from the rail and resolve to this host).
   */
  public Surface withMembers(List<SurfaceRef> newMembers) {
    return new Surface(
        id, presentation, audience, placement, consumes, mountTag, provenance, stateSchema, riskTier,
        altitude, newMembers);
  }

  /**
   * Backwards-compat constructor for callsites that don't yet declare a state schema.
   *
   * <p>Per slice 489 §17.2: stateSchema is optional during the substrate's initial
   * rollout. This constructor mirrors the {@link Operation} backwards-compat
   * constructor pattern: existing 7-arg callsites continue to compile, and the
   * surface registers as "not URL-addressable" (chrome falls through to default state
   * on activation). The conversion to declaring a schema is per-surface, additive.
   */
  public Surface(
      SurfaceRef id,
      Presentation presentation,
      Audience audience,
      Placement placement,
      SurfaceConsumes consumes,
      String mountTag,
      Provenance provenance) {
    this(
        id,
        presentation,
        audience,
        placement,
        consumes,
        mountTag,
        provenance,
        Optional.empty(),
        RiskTier.LOW,
        Altitude.PRODUCT,
        List.of());
  }

  /**
   * Backwards-compat constructor for callsites that declare a state schema but not a
   * {@link RiskTier} (tempdoc 550 WA-4 added the field; existing 8-arg callsites continue to
   * compile and register as {@code RiskTier.LOW} — Navigation to them stays AUTO).
   */
  public Surface(
      SurfaceRef id,
      Presentation presentation,
      Audience audience,
      Placement placement,
      SurfaceConsumes consumes,
      String mountTag,
      Provenance provenance,
      Optional<SurfaceStateSchema> stateSchema) {
    this(
        id,
        presentation,
        audience,
        placement,
        consumes,
        mountTag,
        provenance,
        stateSchema,
        RiskTier.LOW,
        Altitude.PRODUCT,
        List.of());
  }

  /**
   * Backwards-compat constructor for callsites that declare a {@link RiskTier} but not an
   * {@link Altitude} (tempdoc 571 added the field). Existing 9-arg callsites continue to compile and
   * register as {@link Altitude#PRODUCT} — the benign default carrying no foreclosure.
   */
  public Surface(
      SurfaceRef id,
      Presentation presentation,
      Audience audience,
      Placement placement,
      SurfaceConsumes consumes,
      String mountTag,
      Provenance provenance,
      Optional<SurfaceStateSchema> stateSchema,
      RiskTier riskTier) {
    this(
        id,
        presentation,
        audience,
        placement,
        consumes,
        mountTag,
        provenance,
        stateSchema,
        riskTier,
        Altitude.PRODUCT,
        List.of());
  }

  /**
   * Backwards-compat constructor for callsites that declare an {@link Altitude} but not
   * {@link #members} (tempdoc 571 §11 / 578 added the field). Existing 10-arg callsites continue to
   * compile and register as a non-host surface ({@code members = List.of()}).
   */
  public Surface(
      SurfaceRef id,
      Presentation presentation,
      Audience audience,
      Placement placement,
      SurfaceConsumes consumes,
      String mountTag,
      Provenance provenance,
      Optional<SurfaceStateSchema> stateSchema,
      RiskTier riskTier,
      Altitude altitude) {
    this(
        id,
        presentation,
        audience,
        placement,
        consumes,
        mountTag,
        provenance,
        stateSchema,
        riskTier,
        altitude,
        List.of());
  }
}
